File: week-4/score-of-parentheses/score-of-parentheses.js
/**
 * @param {string} S
 * @return {number}
 */
var scoreOfParentheses = function(S) {
    const stack = [];
    let count = 0;
    for (let i of S) {
        if (i === '(') {
            stack.push(count);
            count = 0;
        } else {
            count = stack.pop() + Math.max(count*2, 1);
        }
    }
    
    return count;
};

File: week-4/score-of-parentheses/score-of-parentheses.py
class Solution:
    def scoreOfParentheses(self, S: str) -> int:
        count = 0
        stack = []
        for i in S:
            if i == '(':
                stack.append(count)
                count = 0
            else:
                count = stack.pop() + max(count*2, 1)
        
        return count    
